ml4gw/waveforms/phenom_p.py

```diff
@@ -0,0 +1,731 @@
+                + angcoeffs["epsiloncoeff3"] / omega_cbrt.T
+                + angcoeffs["epsiloncoeff4"] * logomega.T
+                + angcoeffs["epsiloncoeff5"] * omega_cbrt.T
+            )
+            - epsilonoffset
+        ).mT
+
+        cBetah, sBetah = self.WignerdCoefficients(omega_cbrt.T, SL, eta, Sperp)
+
+        cBetah2 = cBetah * cBetah
+        cBetah3 = cBetah2 * cBetah
+        cBetah4 = cBetah3 * cBetah
+        sBetah2 = sBetah * sBetah
+        sBetah3 = sBetah2 * sBetah
+        sBetah4 = sBetah3 * sBetah
+
+        hp_sum = 0
+        hc_sum = 0
+
+        cexp_i_alpha = torch.exp(1j * alpha)
+        cexp_2i_alpha = cexp_i_alpha * cexp_i_alpha
+        cexp_mi_alpha = 1.0 / cexp_i_alpha
+        cexp_m2i_alpha = cexp_mi_alpha * cexp_mi_alpha
+        T2m = (
+            cexp_2i_alpha.T * cBetah4.T * Y2m[0]
+            - cexp_i_alpha.T * 2 * cBetah3.T * sBetah.T * Y2m[1]
+            + 1 * torch.sqrt(torch.tensor(6)) * sBetah2.T * cBetah2.T * Y2m[2]
+            - cexp_mi_alpha.T * 2 * cBetah.T * sBetah3.T * Y2m[3]
+            + cexp_m2i_alpha.T * sBetah4.T * Y2m[4]
+        ).mT
+        Tm2m = (
+            cexp_m2i_alpha.T * sBetah4.T * torch.conj(Y2m[0])
+            + cexp_mi_alpha.T * 2 * cBetah.T * sBetah3.T * torch.conj(Y2m[1])
+            + 1
+            * torch.sqrt(torch.tensor(6))
+            * sBetah2.T
+            * cBetah2.T
+            * torch.conj(Y2m[2])
+            + cexp_i_alpha.T * 2 * cBetah3.T * sBetah.T * torch.conj(Y2m[3])
+            + cexp_2i_alpha.T * cBetah4.T * torch.conj(Y2m[4])
+        ).mT
+        hp_sum = T2m + Tm2m
+        hc_sum = 1j * (T2m - Tm2m)
+
+        eps_phase_hP = torch.exp(-2j * epsilon) * hPhenom / 2.0
+
+        hp = eps_phase_hP * hp_sum
+        hc = eps_phase_hP * hc_sum
+
+        return hp, hc
+
+    def PhenomPOneFrequency(
+        self,
+        fs,
+        m1,
+        m2,
+        eta,
+        eta2,
+        Seta,
+        chi1,
+        chi2,
+        chi12,
+        chi22,
+        chip,
+        phic,
+        M,
+        xi,
+        dist_mpc,
+    ):
+        """
+        m1, m2: in solar masses
+        phic: Orbital phase at peak of the underlying non precessing model
+        M: Total mass (Solar masses)
+        """
+
+        M_s = M * MTSUN_SI
+        Mf = torch.outer(M_s, fs)
+        fRD, _ = self.phP_get_fRD_fdamp(m1, m2, chi1, chi2, chip)
+
+        phase, _ = self.phenom_d_phase(Mf, m1, m2, eta, eta2, chi1, chi2, xi)
+        phase = (phase.T - (phic + PI / 4.0)).mT
+        Amp = self.phenom_d_amp(
+            Mf, m1, m2, eta, eta2, Seta, chi1, chi2, chi12, chi22, xi, dist_mpc
+        )[0]
+        Amp0 = self.get_Amp0(Mf, eta)
+        dist_s = dist_mpc * MPC_SEC
+        Amp = ((Amp0 * Amp).T * (M_s**2.0) / dist_s).mT
+        # phase -= 2. * phic; # line 1316 ???
+        hPhenom = Amp * (torch.exp(-1j * phase))
+
+        linspace = torch.linspace(0.5, 1.5, 101)
```

Comment:
Let's use a variable name that describes the quantity. And like the comment below, this defines the quantity on the CPU which might raise an error when `fRD` is a tensor on the GPU. Have you tested this?

Reply:
I used that to create a grid of frequencies around the ring down frequency for accurate differentiation. It's used immediately in the next line and nowhere else, so I guess it's simpler to just eliminate it and use `torch.linspace` directly.